common/djangoapps/student/docs/decisions/0001-anonymous-user-id-generation.rst

```diff
@@ -0,0 +1,54 @@
+Anonymous User Id Generation
+--------------
+
+Status
+======
+
+Accepted
+
+Context
+=======
+
+The student app provides a mechanism to generate multiple anonymous ids for a
+student.  The anonymous ID can be independent of all courses or it can be
+course specific.  To generate the anonymous ID, we currently hash the user's
+``id`` with the Django ``SECRET_KEY`` and a course key if provided.  The
+mapping between the anonymous ID and user ``id`` are saved in the
+``AnonymousUserID`` table.
+
+As it stands, if the ``SECRET_KEY`` is rotated students would get new anonymous
+IDs starting immediately after rotation.  This can cause downstream issues
+where the IDs are output from the system.  For example, the IDs are in tracking
+data and could be used to track a user's activity through a course for research
+purposes.
+
+Decisions
+=========
+
+Once an anonymous ID is generated for a user in a particular LearningContext
+(either a course or some other unit of learning), it will remain that way even
+if the secret used to generate the ID changes.  For any context where an
+anonymous ID does not already exist, a new ID will be generated using the
+latest ``SECRET_KEY``.
```

Comment:
[puzzlement] I was under the impression that we are decoupling anonymous IDs from using `SECRET_KEY` altogether.  Only for backward compatibility reasons, we want to maintain old generated IDs (which were created with the `SECRET_KEY`). See the acceptance criteria in https://openedx.atlassian.net/browse/ARCHBOM-1645.

Or am I misinterpreting that this ADR is for a transitory phase, followed by another ADR/step that decouples the two?

Reply:
The risk we accepted at parking-lot this morning was limited to the current value of the `SECRET_KEY`. Moving forward, however, the design is to randomly generate future anonymous IDs (there's no reason to use a symmetric key for randomization).

Reply:
See https://github.com/edx/edx-platform/pull/26146 for more details:

There is a danger in generating anonymous IDs because some calls to this function do not save the IDs to the database.  Using a mostly consistent hashing approach is a lot safer.  It still breaks on SECRET_KEY rotation but that should not happen very often.

Remaining coupled to the SECRET_KEY was also a choice we made after looking closer at the problem because adding multiple hashing peppers to the system is expensive and error prone.  Setting a default would result in many openedx instances not remembering to update it. Forcing everyone to set one(which is what Django and edx-platform currently does with `SECRET_KEY`) would result in more changes across the system now as we generate and publish the change everywhere and more work for all openedx operators.

The value of that extra work was questionable to us because, the new secret is sitting next to SECRET_KEY and so has the same security/exposure profile.  If one got leaked, the chances that the other did not is very low.  This means in general we're adding overhead effort of having multiple keys with getting much benefit from it.

Reply:
@feanil Perhaps a real-time conversation would help, in case I'm missing something. I am not supportive of us continuing to use the `SECRET_KEY` for anonymous user ids.  I'd prefer to use Django's symmetric key for temporal purposes, such as session keys - not for persistent identifiers.

If there's a race condition, then we may need to stop creating the anonymous Ids lazily. For example, a user's anonymous id can be created when the user registers. If we need to have course-specific ids, we can create them when enrollments are created.